src/mlpack/methods/ann/layer/multihead_attention_impl.hpp

```diff
@@ -0,0 +1,318 @@
+      Log::Fatal << "The size of the 'attn_mask' is not correct.\n";
+    scores.each_slice() += attnMask;
+  }
+
+  //! Apply the key padding mask when provided. It blacks-out any particular
+  //! word in the sequence.
+  //! The key padding mask has elements 0 or -infinity.
+  //! The shape of keyPaddingMask : (1, srcSeqLen).
+  if (!keyPaddingMask.is_empty())
+  {
+    if (keyPaddingMask.n_rows != 1 || keyPaddingMask.n_cols != srcSeqLen)
+        Log::Fatal << "The size of the 'keyPaddingMask' is not correct.\n";
+    scores.each_slice() += arma::repmat(keyPaddingMask, tgtSeqLen, 1);
+  }
+
+  for (size_t i = 0; i < numHeads * batchSize; ++i)
+  {
+    softmax.Forward(scores.slice(i), softmax.OutputParameter());
+    scores.slice(i) = softmax.OutputParameter();
+  }
+
+  //! Calculate the attention output i.e. matrix multiplication of softmax
+  //! output and vProj.
+  //! The shape of attnOutput : (tgtSeqLen, headDim, numHeads * batchSize).
+  attnOut = math::MultiplyCube2Cube(scores, vProj, false, false);
+
+  //! Now we will concatenate output of all the heads i.e. we will reshape
+  //! attnOut to (tgtSeqLen, embedDim, batchSize).
+  attnOut.reshape(tgtSeqLen, embedDim, batchSize);
+
+  //! The final output is the linear projection of attention output.
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    output.col(i) = arma::vectorise(attnOut.slice(i) * outWt
+        + arma::repmat(outBias, 1, tgtSeqLen).t());
+  }
+}
+
+template <typename InputDataType, typename OutputDataType,
+          typename RegularizerType>
+template <typename eT>
+void MultiheadAttention<InputDataType, OutputDataType, RegularizerType>::
+Backward(const arma::Mat<eT>& /* input */,
+         const arma::Mat<eT>& gy,
+         arma::Mat<eT>& g)
+{
+  typedef typename arma::Cube<eT> CubeType;
+
+  const size_t tgtSeqLen = gy.n_rows / embedDim;
+  const size_t batchSize = gy.n_cols;
+
+  //! Reshape the propagated gradient into a cube.
+  //! The shape of gyTemp : (tgtSeqLen, embedDim, batchSize).
+  //! We need not split it into n heads now because this is the part when
+  //! output were concatenated from n heads.
+  CubeType gyTemp(const_cast<arma::Mat<eT>&>(gy).memptr(), tgtSeqLen, embedDim,
+      batchSize, true, false);
+
+  //! The shape of gyTemp : (tgtSeqLen, embedDim, batchSize).
+  //! The shape of outWt : (embedDim, embedDim).
+  gyTemp = math::MultiplyCube2Mat(gyTemp, outWt, false, true);
+
+  //! Now since the shape of gyTemp is (tgtSeqLen, embedDim, batchSize). We will
+  //! split it into n heads.
+  //! The shape of gyTemp : (tgtSeqLen, headDim, numHeads * batchSize).
+  gyTemp.reshape(tgtSeqLen, headDim, numHeads * batchSize);
+
+  //! The shape of gyTemp : (tgtSeqLen, headDim, numHeads * batchSize).
+  //! The shape of vProj : (srcSeqLen, headDim, numHeads * batchSize).
+  //! So the new shape of gyTemp : (tgtSeqLen, srcSeqLen, numHeads * batchSize).
+  gyTemp = math::MultiplyCube2Cube(gyTemp, vProj, false, true);
```

Comment:
@lozhnikov: Should I remove all such comments? I just mentioned it to understand what is going on?
And can you please explain how memory checks passed? I'm not getting it 😅

Reply:
> Should I remove all such comments?

Leave them as is. Actually, they're very helpful. :+1: 

> And can you please explain how memory checks passed? I'm not getting it.

Neither am I. Perhaps, these issues were fixed in the upstream (you've just merged it).